server/segment_rpc.go

```diff
@@ -816,3 +825,23 @@ func validatePrice(sess *BroadcastSession) error {
 	}
 	return nil
 }
+
+func sendReqWithTimeout(req *http.Request, timeout time.Duration) (*http.Response, error) {
+	type result struct {
+		resp *http.Response
+		err  error
+	}
+	resCh := make(chan result)
+
+	go func() {
+		resp, err := httpClient.Do(req)
+		resCh <- result{resp, err}
+	}()
+
+	select {
+	case res := <-resCh:
+		return res.resp, res.err
+	case <-time.After(timeout):
+		return nil, errTimeout
```

Comment:
This does not stop the request to continue to be sent, and may also create a resource leak in case the request finishes after we return from here. In that case, we would have:
 - 1 stuck goroutine trying to send on the `resCh` with no receiver on the other side. This can be fixed by adding a little buffer (`1`) to the `resCh` above so it doesn't block.
 - 1 `http.Response` object that won't be processed nor have it's body closed for disposal/connection re-use. We could try to fix this in some ways, like for example by not having the channel buffer at all, but instead having `select` case in the goroutine above, which `defaults` to just ignoring disposing the response in case no one is receiving on the other side.

I think both of these issues can be fixed in a different way tho, by inverting the logic of how we handle the cancellation goroutine here: Instead of doing the request in a separate go-routine and only waiting for the result in the current one, create a separate goroutine for handling the timeout/cancellation and do the main work on the main routine executing this function. For the cancellation, we use a custom context that we replace in the request in order to cancel it early (upload timeout) _only if the response headers haven't been received_. Sounds like a fun logic so let me give it a `go`:
```
(loading . . .)
```

Reply:
> What do you mean? Did you mean that calling sendReqWithTimeout() should be in a separate goroutine? I guess there is no need for it.

Ah, no! I just meant that it would be easier to deal with the cancellation, resource disposal and errors if we call `httpClient.Do` directly from `sendReqWithTimeout`, instead of from a goroutine started from it that sent the result over a channel. I was trying to explain the changes that I ended up proposing with the implementation.

Reply:
Great, thanks 👍